### PR TITLE
feat: port addTacticDoc

### PR DESCRIPTION
**Pull request by lakesare**
Ports the `add_tactic_doc` and `getTacticDocEntries` from [Lean 3 doc_commands.lean](https://github.com/leanprover-community/mathlib/blob/master/src/tactic/doc_commands.lean).  
A notable change is Lean 3 version used user attributes to memorize the doc entries, Lean 4 version uses environment extension.

### Questions
- [x] Should we rename `TacticDocEntry` just to `DocEntry`, `add_tactic_doc` to `add_doc_entry`, etc.?
- [x] Rename `inherit_description_from` to `inherit_docstring_from`?
- [x] Whomever will be reviewing this - please take a look at the related PR at https://github.com/leanprover-community/mathlib4/pull/499#discussion_r1023405948, answer to that question will tell us how to name/group our documentation files.

### CODE REVIEW

File: test/AddTacticDoc.lean
Comment (dwrensha) [inline test/AddTacticDoc.lean:34]: Should this be `add_tactic_doc` instead of `add_tactic_doc`?

File: test/AddTacticDoc.lean
Comment (dwrensha) [inline test/AddTacticDoc.lean:43]: Should this be `add_tactic_doc` instead of `add_tactic_dic`?

File: test/AddTacticDoc.lean
Comment (lakesare) [reply test/AddTacticDoc.lean:34]: 😔 yes

**Comment by dwrensha** [comment]
> Should we rename TacticDocEntry just to DocEntry, add_tactic_doc to add_doc_entry, etc.?

My opinion is that it's better to keep the old names for now, to maximize discoverability during the port.  (I don't think that `#align` can help us much, because these things are primarily used in tactic files, which are getting ported by hand.) If we want to rename them before the port is done, then backporting the renaming to mathlib3 first seems to me like way to go.

File: Mathlib/Tactic/AddTacticDoc.lean
Comment (dwrensha) [inline Mathlib/Tactic/AddTacticDoc.lean:84]: It looks like your version actually requires that the `TacticDocEntry` is split out into a separate def, like this:
```lean
def myTde : tde: TacticDocEntry { name := "display name of the tactic",
  category := cat,
  decl_names := [`dcl_1, `dcl_2],
  tags := ["tag_1", "tag_2"] }
  
  add_tactic_doc myTde
```

Is that right? What is the reason for this change from the mathlib3 version?

File: Mathlib/Tactic/AddTacticDoc.lean
Comment (dwrensha) [inline Mathlib/Tactic/AddTacticDoc.lean:120]: This `unsafe` makes me nervous. Is there a way to avoid it?

File: Mathlib/Tactic/AddTacticDoc.lean
Comment (lakesare) [reply Mathlib/Tactic/AddTacticDoc.lean:66]: I used `tacticDocEntry` as "an instance of `TacticDocEntry`", is that not an accepted usage in Lean?

File: Mathlib/Tactic/AddTacticDoc.lean
Comment (digama0) [reply Mathlib/Tactic/AddTacticDoc.lean:120]: This particular pattern is fairly normal, and I would expect it to come up in something like this where we want to make an actual `TacticDocEntry` from a syntax representing such. This is automated to some extent by `declare_config_elab` but that is specialized for config arguments in tactics.

File: Mathlib/Tactic/AddTacticDoc.lean
Comment (lakesare) [reply Mathlib/Tactic/AddTacticDoc.lean:84]: The change is accidental, it should accept a structure!
Do you have some pointers as to what `tdeStx:______` should be to accept a structure?

**Comment by lakesare** [comment]
I removed the `DocCommands.lean` file because we need neither `add_decl_doc` (it's already implemented in Lean 3) nor `copy_doc_string` (we can use `attribute [inherit_doc ffrom] tto`).
Alternatively, we could put `addTacticDoc` command into `DocCommands.lean` and describe the new documentation methods in the module docstring (Lean 3's `add_decl_doc` and `inherit_doc` attribute).

**Comment by dwrensha** [comment]
> I removed the DocCommands.lean file because we need neither add_decl_doc (it's already implemented in Lean 3) nor copy_doc_string (we can use attribute [inherit_doc ffrom] tto).

Note that Mario already removed this file in https://github.com/leanprover-community/mathlib4/pull/684.

**Review by dwrensha — APPROVED** [review]

**Comment by lakesare** [comment]
Should it be merged/are we waiting for the second reviewer, @dwrensha?

**Comment by dwrensha** [comment]
I don't have the permission to merge, so this is waiting on the review of someone who does.

File: test/AddTacticDoc.lean
Comment (digama0) [inline test/AddTacticDoc.lean:57]: Can it also support the following syntax:
```suggestion
add_tactic_doc where
  name       := "never defined separately"
  category   := DocCategory.hole_cmd
  decl_names := [`one]
```
You can implement this as a macro that expands to the version with the { }.

Comment (lakesare) [reply test/AddTacticDoc.lean:57]: It can't!  
Not super sure how to do it though (I'm porting another tactic at the moment, so I might only return to learning macros in a few weeks, wouldn't want this PR to go stale in the meantime).

Something along these lines?

```
macro metaDocString:docComment ? "add_tactic_doc " tdeStx:Lean.Parser.Command.whereStructInst : command => do
  `(add_tactic_doc $tdeStx)
```

Comment (digama0) [reply test/AddTacticDoc.lean:57]: kind of, except that the macro you just wrote calls itself recursively. You would need to convert the where expression into a structure declaration with braces so that you can call the other version of the declaration.

**Comment by jcommelin** [comment]
This pull request has not seen any activity for quite a while, and it does not pass all CI checks or has merge conflicts. For these reasons, I am closing it now.

If you are still interested in getting this PR merged, please re-open the PR, update the branch, and make sure that all CI checks pass, Please label the PR with `awaiting-review` when ready.

**Pull request closed without merging**
